# Day 4 — Pandas: Tabular Data                     

**Goal:** 
● Load a real dataset into a DataFrame and inspect its structure.                  
● Select columns and filter rows by condition.            
● Clean missing values and duplicates, and aggregate data with groupby.                         

---
## Step 1: Loading the dataset
This loads the CSV dataset into a DataFrame and reports its shape, columns, and dtypes.

In [1]:
import pandas as pd
df = pd.read_csv("tested.csv")
print("Dataset Shape:", df.shape) 
print("----------------------")
print("Dataset Info:")
df.info() 


Dataset Shape: (418, 12)
----------------------
Dataset Info:
<class 'pandas.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Survived     418 non-null    int64  
 2   Pclass       418 non-null    int64  
 3   Name         418 non-null    str    
 4   Sex          418 non-null    str    
 5   Age          332 non-null    float64
 6   SibSp        418 non-null    int64  
 7   Parch        418 non-null    int64  
 8   Ticket       418 non-null    str    
 9   Fare         417 non-null    float64
 10  Cabin        91 non-null     str    
 11  Embarked     418 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 39.3 KB


---
## Step 2: Identify Missing Values
Count missing values per column and handle them (fill or drop) with a documented justification

In [2]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

Here we dropped the column Cabin because it had too many missing values (327 missing values filling it would be unoptimized).                  
We filled the whole column Age with the mean of the column (has only 86 missing values it would be best to fill it).                    
we dropped the missing values in the column Fare because it was only 1 missing value. 

In [3]:
df = df.drop(columns=["Cabin"])
df["Age"] = df["Age"].fillna(df["Age"].mean())
df = df.dropna(subset=["Fare"])

In [5]:
df.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

---
## Step 3 + 4: Filter the data to a meaningful subset & Use groupby
Filtered the data to a meaningful subset and described what i found then Used groupby to compute an aggregate statistic per category and interpret the result in a Markdown cell.

In [4]:
adults_df = df[df["Age"] > 18]
print(f"We filtered down to {len(adults_df)} adult passengers.")

adult_count = adults_df.groupby("Sex")["Age"].count()
print("\nAdult Count by Sex:\n", adult_count)


We filtered down to 363 adult passengers.

Adult Count by Sex:
 Sex
female    128
male      235
Name: Age, dtype: int64
